src/main/java/org/broadinstitute/hellbender/engine/AssemblyRegionIterator.java

```diff
@@ -0,0 +1,226 @@
+        this.features = features;
+        this.evaluator = evaluator;
+        this.minRegionSize = minRegionSize;
+        this.maxRegionSize = maxRegionSize;
+        this.assemblyRegionPadding = assemblyRegionPadding;
+        this.activeProbThreshold = activeProbThreshold;
+        this.maxProbPropagationDistance = maxProbPropagationDistance;
+
+        this.readyRegion = null;
+        this.previousRegion = null;
+        this.pendingRegions = new ArrayDeque<>();
+        this.readCachingIterator = new ReadCachingIterator(readShard.iterator());
+        this.readCache = new ArrayDeque<>();
+        this.activityProfile = new BandPassActivityProfile(null, maxProbPropagationDistance, activeProbThreshold, BandPassActivityProfile.MAX_FILTER_SIZE, BandPassActivityProfile.DEFAULT_SIGMA, readHeader);
+
+        // We wrap our LocusIteratorByState inside an AllLocusIterator so that we get empty loci
+        // for uncovered locations. This is critical for reproducing GATK 3.x behavior!
+        final LocusIteratorByState libs = new LocusIteratorByState(readCachingIterator, DownsamplingMethod.NONE, false, ReadUtils.getSamplesFromHeader(readHeader), readHeader, false);
+        this.locusIterator = new AllLocusIterator(readShard.getInterval(), libs);
+
+        readyRegion = loadNextAssemblyRegion();
+    }
+
+    @Override
+    public boolean hasNext() {
+        return readyRegion != null;
+    }
+
+    @Override
+    public AssemblyRegion next() {
+        if ( ! hasNext() ) {
+            throw new NoSuchElementException("next() called when there were no more elements");
+        }
+
+        previousRegion = readyRegion;
+        readyRegion = loadNextAssemblyRegion();
+        return previousRegion;
+    }
+
+    private AssemblyRegion loadNextAssemblyRegion() {
+        AssemblyRegion nextRegion = null;
+
+        while ( locusIterator.hasNext() && nextRegion == null ) {
+            final AlignmentContext pileup = locusIterator.next();
+
+            // Ignore sites that are not contained within this shard's interval
+            if ( ! readShard.getInterval().contains(pileup) ) {
+                continue;
+            }
+
+            // Pop any new pending regions off of the activity profile. These pending regions will not become ready
+            // until we've traversed all the reads that belong in them.
+            //
+            // Ordering matters here: need to check for forceConversion before adding current pileup to the activity profile
+            if ( ! activityProfile.isEmpty() ) {
+                final boolean forceConversion = pileup.getLocation().getStart() != activityProfile.getEnd() + 1;
+                pendingRegions.addAll(activityProfile.popReadyAssemblyRegions(assemblyRegionPadding, minRegionSize, maxRegionSize, forceConversion));
+            }
+
+            // Add the current pileup to the activity profile
+            if ( readShard.getPaddedInterval().contains(pileup.getLocation()) ) {
```

Comment:
Second layer of redundancy which has already been asserted above on line 137

Reply:
Removed